### PR TITLE
Merging code from glowing dangerzone - BLOCKS DEVELOPMENT - please review

### PR DESCRIPTION
**Pull request by josenavas**
This PR is merging the code from glowing dangerzone.

The code in SQLConnectionHandler was a bit hard to follow in some cases because there was no logical structure on the code. I've moved some functions around and github make a fair good job here, it looks like it only had a problem with the `execute_queue` function.

The functions `fetchone_with_types` and `fetchall_with_types` have been removed. They're not used anywhere and nobody knows why they're there.

I've rewritten some of the tests and added a BUNCH of new tests, that is one of the reasons why it shows so many changes.

Once this is merged, I will start implementing all the actual improvements to the SQLConnectionHandler that will require changes in the code base - but that will be done in a different branch.

As a reminder, all this changes are needed to get the portal types in place.

### CODE REVIEW

**Comment by josenavas** [comment]
Ready for review. This is blocking development.

File: qiita_db/sql_connection.py [hunk 1]
```diff
@@ -100,6 +101,18 @@ def flatten(listOfLists):
 
 
 class SQLConnectionHandler(object):
+    """Encapsulates the DB connection with the Postgres DB
```
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: Maybe -> "Postgres DB connection object"?

File: qiita_db/sql_connection.py [hunk 2]
```diff
@@ -100,6 +101,18 @@ def flatten(listOfLists):
...
+
+    Parameters
+    ----------
+    admin : {'no_admin', 'no_database', 'database'}, optional
```
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: These values don't match with the options described in the docstring.

File: qiita_db/sql_connection.py [hunk 3]
```diff
@@ -100,6 +101,18 @@ def flatten(listOfLists):
...
+        Whether or not to connect as the admin user. Options other than
+        `no_admin` depend on admin credentials in the qiita configuration. If
```
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: Minor but you are using a mixture of ` and of ' to refer to the options.

File: qiita_db/sql_connection.py [hunk 4]
```diff
@@ -255,33 +256,30 @@ def get_postgres_cursor(self):
         except PostgresError as e:
             raise QiitaDBConnectionError("Cannot get postgres cursor! %s" % e)
 
-    def set_autocommit(self, on_or_off):
-        """Sets the isolation level to autocommit or default (read committed)
```
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: Would be good to move this docstring to the autocommit property. Otherwise this information is being lost.

File: qiita_db/sql_connection.py [hunk 5]
```diff
@@ -255,33 +256,30 @@ def get_postgres_cursor(self):
...
-        else:
-            level = ISOLATION_LEVEL_READ_COMMITTED
+    @property
+    def autocommit(self):
+        return self._connection.isolation_level == ISOLATION_LEVEL_AUTOCOMMIT
 
+    @autocommit.setter
+    def autocommit(self, value):
+        if not isinstance(value, bool):
+            raise TypeError("The value for autocommit should be a boolean")
```
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: This exception should be documented in the autocommit docstring.

File: qiita_db/sql_connection.py [hunk 6]
```diff
@@ -320,158 +318,55 @@ def _sql_executor(self, sql, sql_args=None, many=False):
...
-            Default False
+            The SQL query
+        sql_args : tuple, list or dict, optional
+            The arguments for the SQL query
 
         Raises
         ------
-        KeyError
-            queue does not exist
+        QiitaDBExecutionError
```
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: I'm OK with this as per our discussions before, but I want to remark (again) that with this we explicitly say that people shouldn't use `SQLConnectionHandler` outside of Qiita.

File: qiita_db/sql_connection.py [hunk 7]
```diff
@@ -320,158 +318,55 @@ def _sql_executor(self, sql, sql_args=None, many=False):
...
+            if there is some error executing the SQL query
 
         Notes
         -----
-        Queues are executed in FIFO order
+        From psycopg2 documentation, only variable values should be bound
```
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: Would be nice to include a `References` section with that link.

File: qiita_db/sql_connection.py [hunk 8]
```diff
@@ -320,158 +318,55 @@ def _sql_executor(self, sql, sql_args=None, many=False):
...
-        else:
-            self._check_sql_args(sql_args)
-            self.queues[queue].append((sql, sql_args))
+        with self._sql_executor(sql, sql_args):
+            pass
 
-    def execute_fetchall(self, sql, sql_args=None):
-        """ Executes a fetchall SQL query
+    def executemany(self, sql, sql_args_list):
+        """ Executes an executemany SQL query with no results
```
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: There's a leading space at the beginning of this docstring.

File: qiita_db/sql_connection.py [hunk 9]
```diff
@@ -479,15 +374,13 @@ def execute_fetchall(self, sql, sql_args=None):
 
         Notes
         -----
-        from psycopg2 documentation, only variable values should be bound
+        From psycopg2 documentation, only variable values should be bound
```
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: Same as above.

File: qiita_db/sql_connection.py [hunk 10]
```diff
@@ -521,22 +414,20 @@ def execute_fetchone(self, sql, sql_args=None):
 
         return result
 
-    def fetchall_with_types(self, sql, sql_args=None):
-        """Executes a fetchall SQL query with column information
+    def execute_fetchall(self, sql, sql_args=None):
+        """ Executes a fetchall SQL query
```
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: There's a leading space in this docstring.

File: qiita_db/sql_connection.py [hunk 11]
```diff
@@ -545,99 +436,156 @@ def fetchall_with_types(self, sql, sql_args=None):
...
-                     for desc in pgcursor.description}
 
-        return result, types
+        return result
 
-    def fetchone_with_types(self, sql, sql_args=None):
-        """Executes a fetchone SQL query with column information
+    def _check_queue_exists(self, queue_name):
+        if queue_name not in self.queues:
+            raise KeyError("Queue %s does not exist" % queue_name)
```
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: For the sake of clarity with whitespaces, etc you may want to do `"Queue '%s' does not exist"`.

File: qiita_db/sql_connection.py [hunk 12]
```diff
@@ -545,99 +436,156 @@ def fetchall_with_types(self, sql, sql_args=None):
...
+        self.queues[queue_name] = []
 
-    def execute(self, sql, sql_args=None):
-        """ Executes an SQL query with no results
+    def list_queues(self):
+        """Returns list of all queue names currently in handler
+
+        Returns
+        -------
+        list
```
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: list of str ... correct?

File: qiita_db/sql_connection.py [hunk 13]
```diff
@@ -545,99 +436,156 @@ def fetchall_with_types(self, sql, sql_args=None):
...
+        for args in sql_args:
+            self._check_sql_args(args)
+            self.queues[queue].append((sql, args))
+
+    def _rollback_raise_error(self, queue, sql, sql_args, e):
+        self._connection.rollback()
+        # wipe out queue since it has an error in it
+        del self.queues[queue]
+        raise QiitaDBExecutionError(
+            ("\nError running SQL query in queue %s: %s\nARGS: %s\nError: %s"
```
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: Why do we need the extra parenthesis here and the leading `'\n'`?

File: qiita_db/sql_connection.py [hunk 14]
```diff
@@ -545,99 +436,156 @@ def fetchall_with_types(self, sql, sql_args=None):
...
-            via sql_args, it shouldn't be used to set table or field names. For
-            those elements, ordinary string formatting should be used before
-            running execute.
+        Queues are executed in FIFO order
         """
-        with self._sql_executor(sql, sql_args_list, True):
-            pass
+        self._check_queue_exists(queue)
+
+        with self.get_postgres_cursor() as cur:
```
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: This is a rather complex block, are we sure that it cannot be decomposed/restructured in any other way?

**Comment by ElDeveloper** [comment]
A few comments, overall looks good and the tests are passing!

File: qiita_db/sql_connection.py (hunk 14 above)
Comment (squirrelo) [inline qiita_db/sql_connection.py]: agree, I can't quite tell whats going on here.

File: qiita_db/sql_connection.py (hunk 1 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Done

File: qiita_db/sql_connection.py (hunk 2 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Good catch. Done

File: qiita_db/sql_connection.py (hunk 3 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Done

File: qiita_db/sql_connection.py (hunk 4 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Yeah, it was not documented on GD and I forgot to move the docs. Thanks!

File: qiita_db/sql_connection.py (hunk 5 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Done

File: qiita_db/sql_connection.py (hunk 6 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: What do you mean by this? Note that the SQLConnectionHandler is part of Qiita, so if they are using this code they are using Qiita. If at some point we actually need this code outside of Qiita, we can then create a separate project. The problem right now is that the SQLConnectionHandler still needs a fair amount of development, and it is hard to maintain the SQLConnectionHandler and Qiita up to date if they're 2 different projects. Once all this code is more or less stable, we can think if this can be move to another project if we need it.

File: qiita_db/sql_connection.py (hunk 7 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Done

File: qiita_db/sql_connection.py (hunk 8 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Done

File: qiita_db/sql_connection.py (hunk 9 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Done

File: qiita_db/sql_connection.py (hunk 10 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Grr sorry about that! Done

File: qiita_db/sql_connection.py (hunk 6 above)
Comment (squirrelo) [inline qiita_db/sql_connection.py]: Similar to MOI, where we try and keep a logical separation even if it is being developed inside qiita.

File: qiita_db/sql_connection.py (hunk 11 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Done - good catch

File: qiita_db/sql_connection.py (hunk 12 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: yup!

File: qiita_db/sql_connection.py (hunk 13 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Good catch! I just move this function around, good to know that reviewing code that has been already reviewed improves it!

File: qiita_db/sql_connection.py (hunk 6 above)
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: @josenavas, yes exactly what you mention is what I was referring to!

Comment (josenavas) [inline qiita_db/sql_connection.py]: @squirrelo MOI is a separate project, so there is a difference.
@ElDeveloper I can remove any kind of QiitaDB errors and use standard python errors so if we actually want to move it to its own project will be easier. What do you think?

File: qiita_db/sql_connection.py (hunk 14 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Yeah... I will try to decompose this. When even its own original developer doesn't know what is going on in the code (@squirrelo) means that the code is really convoluted. Note that I just moved this function around :smile:

File: qiita_db/sql_connection.py (hunk 6 above)
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: I think that this should be a concern, if we think this will be moving
outside of Qiita at some point.

Comment (josenavas) [inline qiita_db/sql_connection.py]: Ok, removing any QiitaDB error and using standard python ones.

File: qiita_db/sql_connection.py (hunk 14 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Thinking a little bit more about this, I need to re-work this block of code in a future PR (which I will start working on it as soon as this one gets merged). Can we leave it as it is acknowledging that is going to disappear soon anyways?

**Comment by josenavas** [comment]
Ready for review again!

**Comment by adamrp** [comment]
@josenavas Looks like there are legitimate failures (something to do with `group` being an SQL reserved term, I think?)

**Comment by josenavas** [comment]
thanks @adamrp 
the issue is that we change some of the errors that are raised in the connection handler, so we use only standard python errors. Other parts of the code rely on those errors to work correctly, so I change those parts of the code to catch the new error types.

I've also checked the interface and I make sure that everywhere that `QiitaDBExecutionError` was catch, `ValueError` is also catch.

**Comment by josenavas** [comment]
BTW, there was a bug in one of the tests, but it was masked by the assertRaises so when we changed the type of the errors from the sql handler I found it and fix it.

File: qiita_db/sql_connection.py (hunk 14 above)
Comment (ElDeveloper) [inline qiita_db/sql_connection.py]: OK

**Comment by ElDeveloper** [comment]
Thanks @josenavas :+1: 

One thing though, with the exceptions that you found that needed to be updated, are there any others where this update needs to happen as well? These places would be untested and thus not showing up as errors when the test suite is run.

**Comment by josenavas** [comment]
I did a grep and search for all of them in the code. I think I covered all of them :-)

**Comment by ElDeveloper** [comment]
fantastic, :+1: :fireworks:

**Comment by josenavas** [comment]
@adamrp, can you review?

File: qiita_db/sql_connection.py [hunk 15]
```diff
@@ -504,140 +410,212 @@ def execute_fetchone(self, sql, sql_args=None):
...
 
         Raises
         ------
-        QiitaDBExecutionError
-            if there is some error executing the SQL query
+        KeyError
+            If the queue `queue_name` does not exist in the handler
+        """
+        if queue_name not in self.queues:
+            raise KeyError("Queue '%s' does not exist" % queue_name)
```
Comment (adamrp) [inline qiita_db/sql_connection.py]: This function seems a little weird to me. It checks whether a queue exists, which I would think returns a boolean. Instead, if it exists, None is returned, and if it doesn't exist, an exception is raised.

File: qiita_db/sql_connection.py [hunk 16]
```diff
@@ -504,140 +410,212 @@ def execute_fetchone(self, sql, sql_args=None):
...
+        ------
+        KeyError
+            Queue name already exists
         """
-        with self._sql_executor(sql, sql_args) as pgcursor:
-            result = pgcursor.fetchone()
-            types = {desc[0]: self.TYPE_CODES[desc[1]]
-                     for desc in pgcursor.description}
+        if queue_name in self.queues:
+            raise KeyError("Queue %s already exists" % queue_name)
```
Comment (adamrp) [inline qiita_db/sql_connection.py]: might as well use the `_check_queue_exists` function here, right?

File: qiita_db/sql_connection.py [hunk 17]
```diff
@@ -504,140 +410,212 @@ def execute_fetchone(self, sql, sql_args=None):
...
+        KeyError
+            queue does not exist
 
-        Note: from psycopg2 documentation, only variable values should be bound
-            via sql_args, it shouldn't be used to set table or field names. For
-            those elements, ordinary string formatting should be used before
-            running execute.
+        Notes
+        -----
+        Queues are executed in FIFO order
```
Comment (adamrp) [inline qiita_db/sql_connection.py]: This note seems unnecessary, since queues are by definition FIFO, but I guess it wouldn't hurt to keep it here

File: qiita_db/sql_connection.py [hunk 18]
```diff
@@ -504,140 +410,212 @@ def execute_fetchone(self, sql, sql_args=None):
...
+            self._check_sql_args(args)
+            self.queues[queue].append((sql, args))
+
+    def _rollback_raise_error(self, queue, sql, sql_args, e):
+        self._connection.rollback()
+        # wipe out queue since it has an error in it
+        del self.queues[queue]
+        raise ValueError(
+            "Error running SQL query in queue %s: %s\nARGS: %s\nError: %s"
+            % (queue, sql, str(sql_args), e))
```
Comment (adamrp) [inline qiita_db/sql_connection.py]: For these exceptions, it might be useful to use [`mogrify`](http://initd.org/psycopg/docs/cursor.html#cursor.mogrify) to get the exact query that it was trying to execute. Not blocking though

File: qiita_db/sql_connection.py [hunk 19]
```diff
@@ -504,140 +410,212 @@ def execute_fetchone(self, sql, sql_args=None):
...
-            The arguments for the SQL query
+        queue : str
+            Name of queue to execute
+
+        Notes
+        -----
+        Does not support executemany command. Instead, enter the multiple
+        SQL commands as multiple entries in the queue.
+
+        Queues are executed in FIFO order
```
Comment (adamrp) [inline qiita_db/sql_connection.py]: same comment about FIFO note here

File: qiita_db/sql_connection.py [hunk 20]
```diff
@@ -504,140 +410,212 @@ def execute_fetchone(self, sql, sql_args=None):
...
         ------
-        QiitaDBExecutionError
-            If there is some error executing the SQL query
-
-        Note: from psycopg2 documentation, only variable values should be bound
-            via sql_args, it shouldn't be used to set table or field names. For
-            those elements, ordinary string formatting should be used before
-            running execute.
+        IndexError
+            If there a placeholder cannot be replaced with a real value
```
Comment (adamrp) [inline qiita_db/sql_connection.py]: I don't understand this explanation; maybe a word was skipped?

**Comment by adamrp** [comment]
I added a few comments, but I don't think any of them should be considered blocking.

File: qiita_db/sql_connection.py (hunk 15 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: I think you're right, semantically your proposal makes more sense that the code that I have. I will change that.

File: qiita_db/sql_connection.py (hunk 16 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Done

File: qiita_db/sql_connection.py (hunk 17 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Removed - easy to maintain

File: qiita_db/sql_connection.py (hunk 18 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: We had mogrify previously in the code and we end up removing it because it was masking some errors, i.e. mogrify was also raising an error so we where not able to check the actual error. I think it is safer to leave as it is, but if you feel strong about it, I can change.

File: qiita_db/sql_connection.py (hunk 19 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Removed

File: qiita_db/sql_connection.py (hunk 18 above)
Comment (adamrp) [inline qiita_db/sql_connection.py]: No, this is fine -- did not remember running into those issues, but I can how that would be problematic. I guess you could wrap mogrify in a try/except (and on exception print what is currently being printed), but it's not a big deal

File: qiita_db/sql_connection.py (hunk 20 above)
Comment (josenavas) [inline qiita_db/sql_connection.py]: Done.

**Comment by josenavas** [comment]
Thanks @adamrp 
I think I've addressed all your comments!

**Comment by adamrp** [comment]
:+1: when tests pass. Thanks @josenavas !

**Comment by josenavas** [comment]
Tests passed!!
